test/c/c_test.go

```diff
@@ -1,23 +1,55 @@
 package c_test
 
 import (
+	"bytes"
+	"context"
+	"fmt"
 	"os"
 	"os/exec"
 	"testing"
 
 	"github.com/bazelbuild/rules_go/go/runfiles"
 	"github.com/stretchr/testify/assert"
+	"github.com/tetratelabs/wazero"
+	"github.com/tetratelabs/wazero/imports/wasi_snapshot_preview1"
 )
 
+type runnerFunc = func(string) ([]byte, error)
+
+func runNative(binary string) ([]byte, error) {
+	return exec.Command(binary).CombinedOutput()
+}
+
+func runWasi(binary string) ([]byte, error) {
+	ctx := context.Background()
+	r := wazero.NewRuntime(ctx)
+	defer r.Close(ctx)
+	buf := &bytes.Buffer{}
+	config := wazero.NewModuleConfig().WithStdout(buf).WithStderr(buf).WithArgs("wasi")
+	wasi_snapshot_preview1.MustInstantiate(ctx, r)
+	bin, err := os.ReadFile(binary)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read guest binary: %v", err)
+	}
+	_, err = r.InstantiateWithConfig(ctx, bin, config)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create instantiate module: %v", err)
+	}
+	return buf.Bytes(), nil
+}
+
 func TestYadda(t *testing.T) {
 	want := os.Getenv("WANT")
 
 	binary, err := runfiles.Rlocation(os.Getenv("BINARY"))
 	if err != nil {
 		t.Fatalf("unable to locate guest binary: %v", err)
 	}
-
-	got, err := exec.Command(binary).CombinedOutput()
+	e := map[string]runnerFunc{
+		"NATIVE": runNative,
+		"WASI":   runWasi,
+	}[os.Getenv("EXECUTOR")]
```

Comment:
a simple case statement will do, no need for closures and this map.

Reply:
Done.